prov/hook/src/hook_ep.c

```diff
@@ -256,32 +246,57 @@ int hook_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
 	return ret;
 }
 
-int hook_endpoint(struct fid_domain *domain, struct fi_info *info,
-		  struct fid_ep **ep, void *context)
+int hook_endpoint_init(struct fid_domain *domain, struct fi_info *info,
+		       struct fid_ep **ep, struct hook_ep *myep, void *context)
 {
 	struct hook_domain *dom = container_of(domain, struct hook_domain, domain);
-	struct hook_ep *myep;
 	struct fid *saved_fid;
 	int ret;
 
-	myep = calloc(1, sizeof *myep);
-	if (!myep)
-		return -FI_ENOMEM;
-
 	saved_fid = info->handle;
 	if (saved_fid) {
 		info->handle = hook_to_hfid(info->handle);
 		if (!info->handle)
 			info->handle = saved_fid;
 	}
 	myep->domain = dom;
+
 	hook_setup_ep(dom->fabric->hclass, &myep->ep, FI_CLASS_EP, context);
+
 	ret = fi_endpoint(dom->hdomain, info, &myep->hep, &myep->ep.fid);
+	info->handle = saved_fid;
+
 	if (ret)
-		free(myep);
-	else
-		*ep = &myep->ep;
+		goto err;
 
-	info->handle = saved_fid;
+	*ep = &myep->ep;
+	return 0;
+err:
+	fi_close(&myep->ep.fid);
+	return ret;
+}
+
+int hook_endpoint(struct fid_domain *domain, struct fi_info *info,
+		  struct fid_ep **ep, void *context)
+{
+	struct hook_domain *dom = container_of(domain, struct hook_domain, domain);
+	struct hook_ep *myep;
+	int ret;
+
+	myep = calloc(1, sizeof *myep);
+	if (!myep)
+		return -FI_ENOMEM;
+
+	ret = hook_endpoint_init(domain, info, ep, myep, context);
+	if (ret)
+		goto err;
+
+	ret = hook_ini_fid(dom->fabric->prov_ctx, &myep->ep.fid);
```

Comment:
If a hook provider intercepts a create/close call, then any specified ini/fini call will not be invoked.  This is desired, but we should probably document that ini/fini calls are not invoked if the app hooks a create/close call.  The app must call ini/fini directly, or perform the desired work.

Reply:
There shouldn't be a need for ini/fini to be invoked if the hook provider intercepts create/close.